dsl/DefaultArgs.cc

```diff
@@ -0,0 +1,168 @@
+#include "ast/Helpers.h"
+#include "common/typecase.h"
+#include "core/GlobalState.h"
+
+using namespace std;
+
+namespace sorbet::dsl {
+
+unique_ptr<ast::Expression> mangleSig(core::Context ctx, unique_ptr<ast::Expression> expr, ast::Expression *param) {
+    auto sig = ast::cast_tree<ast::Send>(expr.get());
+    ENFORCE(sig);
+    ENFORCE(sig->fun == core::Names::sig());
+
+    if (auto kw = ast::cast_tree<ast::KeywordArg>(param)) {
+        param = kw->expr.get();
+    }
+
+    auto ident = ast::cast_tree<ast::UnresolvedIdent>(param);
+    if (!ident) {
+        return ast::MK::EmptyTree();
+    }
+    auto name = ident->name;
+
+    unique_ptr<ast::Expression> retType;
+
+    auto send = ast::cast_tree<ast::Send>(sig->block->body.get());
+    if (!send) {
+        return ast::MK::EmptyTree();
+    }
+
+    while (send != nullptr) {
+        switch (send->fun._id) {
+            case core::Names::params()._id: {
+                if (send->args.size() != 1) {
+                    return ast::MK::EmptyTree();
+                }
+                auto *hash = ast::cast_tree<ast::Hash>(send->args[0].get());
+                if (!hash) {
+                    return ast::MK::EmptyTree();
+                }
+                int i = -1;
+                for (auto &key : hash->keys) {
+                    i++;
+                    auto &value = hash->values[i];
+                    auto lit = ast::cast_tree<ast::Literal>(key.get());
+                    if (lit && lit->isSymbol(ctx)) {
+                        auto symName = lit->asSymbol(ctx);
+                        if (name == symName) {
+                            retType = value->deepCopy();
+                        }
+                    }
+                }
+                break;
+            }
+
+            case core::Names::abstract()._id: {
+                // Don't make this method at all since abstract methods can't
+                // have bodies
+                return nullptr;
+            }
+
+            case core::Names::override_()._id: {
+                // A total hack but we allow .void.void or .void.returns and
+                // the one with content wins
+                send->fun = core::Names::void_();
+            }
+        }
+        auto recv = ast::cast_tree<ast::Send>(send->recv.get());
+        send = recv;
+    }
+
+    send = ast::cast_tree<ast::Send>(sig->block->body.get());
+    while (send != nullptr) {
+        switch (send->fun._id) {
+            case core::Names::returns()._id: {
+                if (!retType) {
+                    return ast::MK::EmptyTree();
+                }
+                send->args[0] = move(retType);
+                break;
+            }
+
+            case core::Names::void_()._id: {
+                if (!retType) {
+                    return ast::MK::EmptyTree();
+                }
+                send->fun = core::Names::returns();
+                send->args.emplace_back(move(retType));
+                break;
```

Comment:
These two cases are the same if you always set the name to returns, why not collapse them so that there's less bug surface area?

Reply:
Never mind, on a second read through it feels like that would make the args handling unnecessarily complicated.